fzf-tab.zsh

```diff
@@ -308,24 +308,29 @@ _fzf_tab_complete() {
 
 zle -C _fzf_tab_complete complete-word _fzf_tab_complete
 
+# note: this function is called with user options
 _fzf_tab_try_custom_completion() {
     # do not steal fzf's completions
-    [[ $LBUFFER =~ ${(q)FZF_COMPLETION_TRIGGER-'**'}$ ]] && return 1
-    local tokens=(${(z)LBUFFER})
-    [[ ${LBUFFER[-1]} = ' ' ]] && tokens+=("")
-    local cmd=${tokens[1]}
-    if (( $+functions[${FZF_TAB_CUSTOM_COMPLETIONS_PREFIX}${cmd}] )); then
-        local prefix=${tokens[-1]}
-        local lbuf
-        [ -z "${tokens[-1]}" ] && lbuf=$LBUFFER || lbuf=${LBUFFER:0:-${#tokens[-1]}}
-        prefix="$prefix" eval _fzf_complete_${cmd} ${(q)lbuf}
-        return 0
-    fi
-    return 1
+    [[ $LBUFFER != *"${FZF_COMPLETION_TRIGGER-**}" ]] || return
```

Comment:
`_fzf_tab_orig_widget` should be checked (if it is not fzf-completion, checking FZF_COMPLETION_TRIGGER is pointless, previously this worked because of `q` bug).
```
[[ $_fzf_tab_orig_widget != 'fzf-completion' || $LBUFFER != *"${FZF_COMPLETION_TRIGGER-**}" ]] || return
```

You can quickly check completions of foo in this settings
```
export FZF_COMPLETION_TRIGGER=""
bindkey '^T' fzf-completion
bindkey '^I' fzf-tab-complete
export fzf_default_completion=expand-or-complete
export _fzf_tab_orig_widget=expand-or-complete
_fzf_complete_foo() { _fzf_complete "+m -n 1" "$@" < <(echo bar; echo baz) }
```

Reply:
What you want to check for is whether `zle $_fzf_tab_orig_widget` recognizes `FZF_COMPLETION_TRIGGER`. There is no way to implement this check in the code. Zle widgets get renamed and wrapped left and right. A widget called `expand-or-complete` can very well respect `FZF_COMPLETION_TRIGGER`, while a widget called `fzf-completion` may ignore it.

I'm having mixed feelings about the introduction of custom completions to fzf-tab. fzf-tab used to be focused on doing one thing well -- using fzf to display completion candidates from from the builtin Zsh completion system. Now it also has its own system for generating completion candidates. This poses many problems:

1. How to decide which system to use to generate completions? This is what has started the discussion of `FZF_COMPLETION_TRIGGER`.

2. The alternative completion system has some advantages (e.g., completion candidates can be streamed into fzf instead of accumulating all candidates before firing up fzf) but it also has many disadvantages. The biggest disadvantages is that it doesn't understand zsh syntax.

Users who happened to be using fzf-tab and fzf will now find that completions for several commands (`export`, `unset`, `ssh`, etc.) don't work as well as they used to. For example:

```zsh
unset foo; cd ./<TAB>
```

This used to work. Now it doesn't because it'll trigger custom completion for `unset`.

Perhaps fzf-tab shouldn't provide this extension mechanism until we better understand how it should work? Users who want it can implement their own dispatching widget and bind it to TAB. The dispatching widget can call `fzf-tab-complete`, `fzf-complete` or whatever, based on conditions preferred by the user.

WDYT?

Reply:
> unset foo; cd ./<TAB>

true, did not consider all cases.

>Perhaps fzf-tab shouldn't provide this extension mechanism until we better understand how it should work?

No problem with me.

Reply:
@Aloxaf I propose to merge this PR as it makes code strictly better by fixing a few bugs, and then revert custom completions. Hopefully there aren't many users who rely on this feature yet (just one?) and they can have it by adding a few lines of code to their own configs. These lines can be based on the code from fzf-tab that is going to be reverted, so it would be nice if bug fixes from this PR landed first.

Reply:
No problem. Thanks for both of your impressive discussions.